File: src/server/services/honeypot.service.ts
import { appendAuditBlock, readDb, writeDb } from '@/server/db/database';
import { HoneypotProfile, TwinSyncMetadata } from '@/shared/types';
import { execSync } from 'child_process';
import fs from 'fs';
import os from 'os';
import path from 'path';

export interface DockerStatus {
  available: boolean;
  activeContainersCount: number;
  containers: Array<{ id: string; name: string; status: string; ports: string }>;
}

export async function checkDockerDaemonStatus(): Promise<DockerStatus> {
  try {
    const stdout = execSync('docker ps --format "{{.ID}}|{{.Names}}|{{.Status}}|{{.Ports}}"', {
      timeout: 3000,
      encoding: 'utf-8',
      stdio: ['pipe', 'pipe', 'ignore'],
    });

    const lines = stdout.trim().split('\n').filter(Boolean);
    const containers = lines.map((line) => {
      const [id, name, status, ports] = line.split('|');
      return { id: id || 'unknown', name: name || 'unknown', status: status || 'running', ports: ports || '' };
    });

    return {
      available: true,
      activeContainersCount: containers.length,
      containers,
    };
  } catch {
    return {
      available: false,
      activeContainersCount: 0,
      containers: [],
    };
  }
}

export async function getAllHoneypots(): Promise<HoneypotProfile[]> {
  const db = readDb();
  return db.honeypots;
}

export async function createHoneypot(data: {
  name: string;
  type: 'Cowrie' | 'Dionaea' | 'CustomLLM';
  port: number;
  temporalJitterMs: number;
  twinSyncEnabled: boolean;
}): Promise<HoneypotProfile> {
  const db = readDb();
  const containerId = `doc-${Math.random().toString(36).substring(2, 10)}`;


  try {
    execSync(`docker run -d --name cipher-${containerId} -p ${data.port}:${data.port} alpine sleep infinity`, {
      timeout: 4000,
      stdio: 'ignore',
    });
  } catch {

  }

  const newHp: HoneypotProfile = {
    id: `hp-${data.type.toLowerCase()}-${Date.now().toString(36)}`,
    name: data.name,
    type: data.type,
    status: 'active',
    port: data.port,
    ip: '127.0.0.1',
    containerId,
    twinSyncEnabled: data.twinSyncEnabled,
    temporalJitterMs: data.temporalJitterMs,
    activeSessionsCount: 0,
    totalEventsCount: 0,
    createdAt: new Date().toISOString(),
  };

  db.honeypots.push(newHp);
  writeDb(db);


  appendAuditBlock('HONEYPOT_DECOY_CREATED', { id: newHp.id, type: newHp.type, port: newHp.port });

  return newHp;
}

export async function toggleHoneypotStatus(id: string): Promise<HoneypotProfile | null> {
  const db = readDb();
  const hp = db.honeypots.find((h) => h.id === id);
  if (!hp) return null;

  const nextStatus = hp.status === 'active' ? 'stopped' : 'active';
  hp.status = nextStatus;


  try {
    if (nextStatus === 'stopped') {
      execSync(`docker stop cipher-${hp.containerId}`, { timeout: 3000, stdio: 'ignore' });
    } else {
      execSync(`docker start cipher-${hp.containerId}`, { timeout: 3000, stdio: 'ignore' });
    }
  } catch {

  }

  writeDb(db);
  appendAuditBlock('HONEYPOT_STATUS_TOGGLED', { id: hp.id, newStatus: nextStatus });

  return hp;
}

export async function getDigitalTwinMetadata(): Promise<TwinSyncMetadata> {
  const hostname = os.hostname() || 'ciphernest-node';
  const osRelease = `${os.type()} ${os.release()}`;
  const architecture = os.arch();

  const interfaces = os.networkInterfaces();
  const activeIfaces = Object.keys(interfaces).join(', ');

  const scannedDirs: string[] = [];
  try {
    const rootItems = fs.readdirSync(process.cwd());
    for (const item of rootItems) {
      if (!item.startsWith('.')) {
        const full = path.join(process.cwd(), item);
        try {
          if (fs.statSync(full).isDirectory()) {
            scannedDirs.push(item);
          }
        } catch {

        }
      }
    }
  } catch {
    scannedDirs.push('src', 'public', 'node_modules', 'config');
  }

  return {
    hostname,
    osRelease,
    architecture,
    activePortRange: `2222-2225 (Ifaces: ${activeIfaces.substring(0, 30)})`,
    directoryNaming: Array.from(new Set(scannedDirs)).slice(0, 8),
    filePatterns: ['*.env', '*.config.json', 'package.json', 'tsconfig.json', 'salary_review.csv'],
    lastSyncedAt: new Date().toISOString(),
    syncApproved: true,
  };
}

File: backend/api/services/honeypot_service.py
"""Honeypot service — port of src/server/services/honeypot.service.ts"""
import os, secrets, subprocess, platform
from datetime import datetime, timezone
from ..db import read_db, write_db, append_audit_block

def _now(): return datetime.now(timezone.utc).isoformat()

def check_docker_status() -> dict:
    try:
        out = subprocess.check_output(
            ['docker','ps','--format','{{.ID}}|{{.Names}}|{{.Status}}|{{.Ports}}'],
            timeout=3, stderr=subprocess.DEVNULL
        ).decode()
        containers = []
        for line in out.strip().splitlines():
            if not line: continue
            parts = (line + "|||").split("|")
            containers.append({"id": parts[0], "name": parts[1], "status": parts[2], "ports": parts[3]})
        return {"available": True, "activeContainersCount": len(containers), "containers": containers}
    except Exception:
        return {"available": False, "activeContainersCount": 0, "containers": []}

def get_all_honeypots() -> list:
    return read_db().get("honeypots", [])

def create_honeypot(data: dict) -> dict:
    db = read_db()
    container_id = secrets.token_hex(8)
    started = False
    try:
        subprocess.run(
            ['docker','run','-d','--name',f'cipher-{container_id}',
             '-p',f'{data["port"]}:{data["port"]}',
             '--label','ciphernest.honeypot=true',
             'ciphernest-honeypot-01'],  # built from docker/Dockerfile.honeypot
            timeout=4, check=True, stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL
        )
        started = True
    except Exception:
        pass
    hp = {
        "id": f"hp-{data['type'].lower()}-{int(__import__('time').time()*1000):x}",
        "name": data["name"], "type": data["type"],
        "status": "active" if started else "error",
        "port": data["port"], "ip": "127.0.0.1",
        "containerId": container_id,
        "twinSyncEnabled": data.get("twinSyncEnabled", False),
        "temporalJitterMs": data.get("temporalJitterMs", 200),
        "activeSessionsCount": 0, "totalEventsCount": 0,
        "createdAt": _now(),
    }
    honeypots = db.get("honeypots", [])
    honeypots.append(hp)
    db["honeypots"] = honeypots
    write_db(db)
    append_audit_block("HONEYPOT_DECOY_CREATED", {"id": hp["id"], "type": hp["type"], "port": hp["port"]})
    return hp

def toggle_honeypot(hp_id: str):
    db = read_db()
    hp = next((h for h in db.get("honeypots", []) if h["id"] == hp_id), None)
    if not hp: return None
    new_status = "stopped" if hp["status"] == "active" else "active"
    hp["status"] = new_status
    try:
        cmd = "stop" if new_status == "stopped" else "start"
        subprocess.run(['docker', cmd, f'cipher-{hp["containerId"]}'],
                       timeout=3, stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL)
    except Exception:
        pass
    write_db(db)
    append_audit_block("HONEYPOT_STATUS_TOGGLED", {"id": hp["id"], "newStatus": new_status})
    return hp

def get_digital_twin_metadata() -> dict:
    import socket, os as _os
    hostname = socket.gethostname()
    os_info = f"{platform.system()} {platform.release()}"
    arch = platform.machine()
    import psutil
    ifaces = list(psutil.net_if_addrs().keys()) if hasattr(psutil, "net_if_addrs") else ["lo", "eth0"]
    cwd = _os.getcwd()
    dirs = []
    try:
        for item in _os.listdir(cwd):
            if not item.startswith("."):
                full = _os.path.join(cwd, item)
                if _os.path.isdir(full):
                    dirs.append(item)
    except Exception:
        dirs = ["src", "public", "backend", "data"]
    return {
        "hostname": hostname,
        "osRelease": os_info,
        "architecture": arch,
        "activePortRange": f"2222-2225 (Ifaces: {', '.join(ifaces)[:30]})",
        "directoryNaming": list(set(dirs))[:8],
        "filePatterns": ["*.env", "*.config.json", "package.json", "tsconfig.json"],
        "lastSyncedAt": _now(),
        "syncApproved": True,
    }
